### PR TITLE
scaffolder(next): Support `async` validation for `CustomFieldExtensions` and run validation for `type: object`

### LINKED ISSUE #7868: Support async validation in scaffolder custom field extensions
I think it can be useful to have a possibility to add backend validation in scaffolder form

## Feature Suggestion
Async validation in scaffolder form

## Possible Implementation
Allow async validation in `createScaffolderFieldExtension` or maybe pass `addError` method as prop to field component

### LINKED ISSUE #8420: Validation not called for custom FieldExtension with type object when using v1beta3 template api
## Expected Behavior

Custom validation should be called for custom FieldExtension with type object

## Current Behavior

Custom validation was not called for custom FieldExtension with type object

## Possible Solution

Not a solution but some investigation

The fields show up ok but validation is never called when I move next on the form, I had a look at the code https://github.com/backstage/backstage/blob/master/plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx#L81
Seems like it will never reach to the part where it will call the validator for the `ui-field` as the result will be considered object at if `(isObject(propData)) {` on line https://github.com/backstage/backstage/blob/master/plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx#L81

and then would not go further because of `if (!isObject(schemaProps)) {`  https://github.com/backstage/backstage/blob/master/plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx#L74, I tried to get around it but providing properties like below but then it tries to do their schema validation instead of the ui-field custom validation
```
  parameters:
    - title: Provide Component Details
      required:
        - repoInfo
        - description
      properties:
        repoInfo:
          title: Unique name of the component
          type: object
          ui:field: RepoUrlPickerCustom
          properties:
            workspace:
              type: string
              title: workspace
              description: The workspace
            project:
              type: string
              title: project
              description: The project
            repo:
              type: string
              title: repo
              description: The repo
```

## Steps to Reproduce
https://backstage.io/docs/features/software-templates/writing-custom-field-extensions

1. Create a custom FieldExtension `RepoUrlPickerCustom`, a copy of `RepoUrlPicker`
Use the object type in FieldProps
```
export const RepoUrlPickerCustom = ({
  onChange,
  rawErrors,
  formData,
}: FieldProps<{ workspace: string, project: string, repo: string }>) => {
```
2. Create a custom validator `RepoUrlPickerCustomValidation`, a copy of `validation` for `RepoUrlPicker`
3. Register the FieldExtension
```
export const RepoUrlPickerCustomFieldExtension = plugin.provide(
    createScaffolderFieldExtension({
        name: 'RepoUrlPickerCustomExtension',
        component: RepoUrlPickerCustom,
        validation: RepoUrlPickerCustomValidation,
    }),
);
```
4. Add parameter with `ui:field` in the template
```
  parameters:
    - title: Provide Component Details
      required:
        - repoInfo
        - description
      properties:
        repoInfo:
          title: Unique name of the component
          type: object
          ui:field: RepoUrlPickerCustom
```

5. In the frontend template form, click next without adding anything for fields in `RepoUrlPickerCustom`. The form moves to next stage instead of calling the `RepoUrlPickerCustomValidation`

## Context
Trying to create a custom UI component in the form which can have multiple fields depend on each other and perform validation accordingly, and be able to use the object for multiple values instead of a single string.

## Your Environment

- NodeJS Version (v14): v14.18.1
- Operating System and Version (e.g. Ubuntu 14.04): Windows 10 20H2 ( OS Build 19042.1288)
- Browser Information: Google Chrome - Version 96.0.4664.45 (Official Build) (64-bit)

### PR DESCRIPTION
**Pull request by benjdlambert**
Fixes #7868
Fixes #8420

### CODE REVIEW

File: plugins/scaffolder/src/next/TemplateWizardPage/Stepper/createAsyncValidators.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ */
+
+import { FieldValidation } from '@rjsf/core';
+import { JsonObject } from '@backstage/types';
+import { ApiHolder } from '@backstage/core-plugin-api';
+import { CustomFieldValidator } from '../../../extensions';
+import { Draft07 as JSONSchema } from 'json-schema-library';
+import { createFieldValidation } from './schema';
+
+export const createAsyncValidator = (
```
Comment (freben) [inline plugins/scaffolder/src/next/TemplateWizardPage/Stepper/createAsyncValidators.ts:24]: this file is named with a plural s on the end, but the function isn't

File: plugins/scaffolder/src/next/TemplateWizardPage/Stepper/schema.ts [hunk 2]
```diff
@@ -110,3 +110,19 @@ export const extractSchemaFromStep = (
   extractUiSchema(returnSchema, uiSchema);
   return { uiSchema, schema: returnSchema };
 };
+
+/**
+ * @alpha
+ * Creates a field validation object for use in react jsonschema form
+ * @returns {FieldValidation} A field validation object that can be used to validate a field
```
Comment (freben) [inline plugins/scaffolder/src/next/TemplateWizardPage/Stepper/schema.ts:117]: we should avoid types in docstrings right?

File: plugins/scaffolder/src/next/TemplateWizardPage/Stepper/Stepper.tsx [hunk 3]
```diff
@@ -61,12 +68,38 @@ export const Stepper = (props: StepperProps) => {
...
+  }, [steps, activeStep, validators, apiHolder]);
+
   const handleBack = () => {
     setActiveStep(prevActiveStep => prevActiveStep - 1);
   };
 
-  const handleNext = ({ formData }: { formData: JsonObject }) => {
-    setActiveStep(prevActiveStep => prevActiveStep + 1);
+  const handleNext = async ({ formData }: { formData: JsonObject }) => {
+    setErrors(undefined);
```
Comment (freben) [inline plugins/scaffolder/src/next/TemplateWizardPage/Stepper/Stepper.tsx:89]: probably not already here, right? should it just mark the form as "busy" (greyed out and whatnot) and then update the errors as needed at the end?

File: plugins/scaffolder/src/next/TemplateWizardPage/Stepper/createAsyncValidators.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      const definitionInSchema = parsedSchema.getSchema(
+        `${pathPrefix}/${key}`,
+        formData,
+      );
+
+      if (definitionInSchema && 'ui:field' in definitionInSchema) {
+        const validator = validators[definitionInSchema['ui:field']];
+        if (validator) {
+          const fieldValidation = createFieldValidation();
+          await validator(value, fieldValidation, context);
```
Comment (freben) [inline plugins/scaffolder/src/next/TemplateWizardPage/Stepper/createAsyncValidators.ts:44]: This should have error handling (try/catch around the call).

File: plugins/scaffolder/src/next/TemplateWizardPage/Stepper/Stepper.tsx (hunk 3 above)
Comment (benjdlambert) [reply plugins/scaffolder/src/next/TemplateWizardPage/Stepper/Stepper.tsx:89]: Yeah loading states are a little unsettled yet, as maybe each component could have it's own loading state. Leaving that for now, but yes 💯

**Review by freben — APPROVED** [review]

**Pull request merged**